Cookbook get resources (#57)
mulesoft/docs-munit: modules/ROOT/pages/mock-file-cookbook.adoc

> = Mocking Resources for Your Tests

Reviewer: Do we have to use "resources," which is kind of vague? 

Is the audience actually always just mocking code?
If so, how about Mocking Code for Your Tests, or Mocking Test Code?

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> When the tests suites start to grow in size and complexity, it’s a good practice to externalize some parts of the code to test resources so that you can keep the code clean and organized as well as reuse the files that you need to mock for your tests without the need to have duplicated code.

Reviewer: Will your audience readily know what a "test suite" is?

Reviewer: ```suggestion
As your tests suites start to get bigger and more complex, it’s a good practice to externalize some parts of the test code, not only so that you can keep the code clean and organized but also so that you can reuse the files that you need to mock for your tests.
```

> In Mule 3, you used to use a `getResourceAsString()` function to incorporate external or mocked resources to your tests. In Mule 4, you can leverage DataWeave to mock your code for your tests.

Reviewer: ```suggestion
Mule 3 uses a `getResourceAsString()` function to incorporate external or mocked resources into your tests. In Mule 4, you can leverage DataWeave to mock your test code, either by creating DataWeave scripts or by using variables in a DataWeave module.
```
Adding this text prepares the audience for the two subtopics that follow.

== Mocking Files Using DataWeave Scripts

> You can use a DataWeave script that returns the data that you want to mock, save the file under a folder in `src/test/resources` and then use a DataWeave function to read this file.

Reviewer: QUESTION: 
You use a DreamWeave script that returns the data to be mocked, or the actual mocked data? 
Do you have to write this script? That user action seems missing.
Is the script the same thing as "the file" that saved in the folder?
Is the DataWeave "function" the same as the "script"?

Reviewer 2: You are right it was confusing. Script is not the right word here. I'm sticking with Dataweave file.

> For example, define a `mockPost.dwl` file in the `src/test/resources/sample_data` folder with the following content:

Reviewer: ```suggestion
For example, imagine that you have defined a `mockPost.dwl` file in the `src/test/resources/sample_data` folder with the following content:
```

[source]
----
%dw 2.0
output application/json
---
{
	"foo" : "var"
}
----

> This DataWeave file is a sample that creates a mock payload that you want to return in an `http:request`. When configuring the `then-return` section in the `mock-when`, use the `readUrl` function to read the mapping file:

Reviewer: Does the first sentence here refer to the syntax above, and the second sentence to the syntax below?
If so, I would separate the two sentences as their own paragraphs...maybe even specify in the first sentence:
```suggestion
This DataWeave file (`mockPost.dwl`) is a sample that creates a mock payload that you want to return in an `http:request` request. 

When configuring the `then-return` section in the `mock-when`, use the `readUrl` function to read the mapping file:
```
Will the audience understand the context of the second sentence, given that these syntax elements are not shown until after it?

[source,xml,linenums]
----
<munit-tools:then-return>
  <munit-tools:payload value="#[readUrl('classpath://sample_data/mockPost.dwl')]" mediaType="application/json" encoding="UTF-8" />
</munit-tools:then-return>
----

> The `classpath-based` URL uses the `classpath:` protocol prefix to find the file we defined under `src/test/resources`.

Reviewer: ```suggestion
The `classpath-based` URL uses the `classpath:` protocol prefix to find the file defined in `src/test/resources`.
```

== Mocking Files Using Variables in a DataWeave Module

> You can create a DataWeave module to centralize all your mocks in a single file by declaring them in DataWeave variables.

Reviewer: Here and at the beginning of the script text, consider adding a bit about WHY a user might choose using scripts vs. variables in modules.

Reviewer: ```suggestion
You can create a DataWeave module to centralize all your mocked data in a single file by declaring it using DataWeave variables.
```

> For example, to create a custom DataWeave module, create a `TestData.dwl` file under `src/test/resources` and add a couple of variables with references to different mocks either from files or inline:

Reviewer: ```suggestion
For example, imagine that you create a custom DataWeave module by creating a `TestData.dwl` file in `src/test/resources`, and you add a couple of variables with references to different mocked data, either from files or inline:
```

[source]
----
import getResourceAsString from MunitTools

var mockPost = readUrl('classpath://sample_data/mockPost.dwl')
var foo = "N123"
var jsonFromFile = read(getResourceAsString('sample_data/mydata.json'), 'application/json')
var jsonObject = { paymentID: "1B56925769601335TLQMIWVY" }
----

> You can then access any of the defined variables in the DataWeave module by importing the variables from your custom module:

Reviewer: Does this text refer to one module or two; are "the DataWeave module" and "your custom module" the same?
Clarify, if you think that your audience might be confused (because I am...but I am not as technical as they...).

[source,xml,linenums]
----
<munit-tools:then-return >
				<munit-tools:variables>
        				<munit-tools:variable key="ticket" value="#[import TestData output application/json --- TestData::mockPost]" mediaType="application/json" />
				</munit-tools:variables>
</munit-tools:then-return>
----

> Using this approach with a proper file naming convention and folder structure should help you have cleaner code and organized mocks for your MUnit test suites as well as provide you an easy way to reuse code.

Reviewer: Does this sentence refer to the module approach or both the script and module approach?

Reviewer: Also...is it even necessary, given that the topic is introduced by stating these purposes? Why repeat them?

Reviewer: If you keep this text...:
```suggestion
Using this approach, along with proper file naming and folder structure, helps you have cleaner code and organized mocked data for your MUnit test suites, and enables you to reuse code.
```
Specify what "this approach" is: script, module, or both? Replace with just "DreamWeave" or "DreamWeave functionality"?
